subprojects/snapshots/src/main/java/org/gradle/internal/fingerprint/impl/MissingRootIgnoringFileSystemSnapshotVisitor.java

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.fingerprint.impl;
+
+import org.gradle.internal.file.FileType;
+import org.gradle.internal.snapshot.FileSystemLocationSnapshot;
+import org.gradle.internal.snapshot.RootTrackingFileSystemSnapshotHierarchyVisitor;
+import org.gradle.internal.snapshot.SnapshotVisitResult;
+
+import java.util.HashSet;
+
+/**
+ * A {@link RootTrackingFileSystemSnapshotHierarchyVisitor} that ignores missing roots and multiple entries for the same path.
+ */
+public abstract class MissingRootIgnoringFileSystemSnapshotVisitor extends RootTrackingFileSystemSnapshotHierarchyVisitor {
```

Comment:
Should this be called `MissingRootAndDuplicateIgnoringFileSystemSnapshotVisitor`?

Reply:
I wonder if it would make sense to split the visitor into two, one filtering already visited entries, the other filtering missing roots.

Reply:
I renamed the type to `MissingRootAndDuplicateIgnoringFileSystemSnapshotVisitor`. I don't think it makes sense to split it in two.